src/io/network/addrinfo.hpp

```diff
@@ -11,21 +11,39 @@
 
 #pragma once
 
+#include <netdb.h>
+#include <string>
+
 namespace io::network {
 
+struct Endpoint;
+
 /**
  * Wrapper class for getaddrinfo.
  * see: man 3 getaddrinfo
  */
 class AddrInfo {
-  explicit AddrInfo(struct addrinfo *info);
-
  public:
+  struct Iterator {
+    explicit Iterator(addrinfo *p) : ptr(p) {}
+    addrinfo &operator*() const { return *ptr; }
+    Iterator &operator++() {
+      ptr = ptr->ai_next;
+      return *this;
+    }
+    bool operator!=(const Iterator &rhs) { return rhs.ptr != ptr; };
+
+   private:
+    addrinfo *ptr;
+  };
+
   ~AddrInfo();
 
-  static AddrInfo Get(const char *addr, const char *port);
+  AddrInfo(const std::string &addr, uint16_t port);
+  explicit AddrInfo(const Endpoint &endpoint);
 
-  operator struct addrinfo *();
+  auto begin() const { return Iterator(info); }
+  auto end() const { return Iterator(nullptr); }
```

Comment:
Once you exhaust  `ptr->as_next`, is it guranteed by the implementation that it will point to `nulltpr`? Also, isn't  the case that `end` should always return the address after the last element? I am not sure what the C++ standard mandates but if  it does, then we should follow that.

Reply:
I guess in a linked list you have to set `next` to `nullptr` to mark the end of the list.

I think the standard doesn't require an exact address that is after the last element, but more likely it meant in a way that the end operator should be exclusive.

Reply:
Yes, getaddrinfo() returns the info as a single linked list, using `ai_next` as the pointer to the next address element in the list or `null` in case there are no more list entries, thus the `nullptr` in `end()`.